ethcore/src/miner/price_info.rs

```diff
@@ -15,94 +15,95 @@
 use std::str::FromStr;
-use std::sync::mpsc;
-use hyper::client::{Handler, Request, Response, Client};
-use hyper::{Url, Next, Encoder, Decoder};
-use hyper::net::HttpStream;
+
+use fetch;
+use fetch::{Client as FetchClient, Fetch};
+use futures::Future;
 
 #[derive(Debug)]
 pub struct PriceInfo {
 	pub ethusd: f32,
 }
 
-pub struct SetPriceHandler<F> {
-	set_price: F,
-	channel: mpsc::Sender<()>,
+/// Price info error.
+#[derive(Debug, Clone)]
+pub enum Error {
+	/// The API returned an unexpected status code or content.
+	UnexpectedResponse(&'static str, String),
+	/// There was an error when trying to reach the API.
+	Fetch(String),
+	/// IO error when reading API response.
+	Io(String),
 }
 
-impl<F> Drop for SetPriceHandler<F> {
-	fn drop(&mut self) {
-		let _ = self.channel.send(());
-	}
+impl From<io::Error> for Error {
+	fn from(err: io::Error) -> Self { Error::Io(format!("{:?}", err)) }
 }
 
-impl<F: Fn(PriceInfo) + Sync + Send + 'static> Handler<HttpStream> for SetPriceHandler<F> {
-	fn on_request(&mut self, _: &mut Request) -> Next { Next::read().timeout(Duration::from_secs(3)) }
-	fn on_request_writable(&mut self, _: &mut Encoder<HttpStream>) -> Next { Next::read().timeout(Duration::from_secs(3)) }
-	fn on_response(&mut self, _: Response) -> Next { Next::read().timeout(Duration::from_secs(3)) }
-
-	fn on_response_readable(&mut self, r: &mut Decoder<HttpStream>) -> Next {
-		let mut body = String::new();
-		let info = r.read_to_string(&mut body)
-			.map_err(|e| format!("Unable to read response: {:?}", e))
-			.and_then(|_| self.process_response(&body));
-
-		if let Err(e) = info {
-			warn!("Failed to auto-update latest ETH price: {:?}", e);
-		}
-		Next::end()
-	}
+impl From<fetch::Error> for Error {
+	fn from(err: fetch::Error) -> Self { Error::Fetch(format!("{:?}", err)) }
 }
 
-impl<F: Fn(PriceInfo) + Sync + Send + 'static> SetPriceHandler<F> {
-	fn process_response(&self, body: &str) -> Result<(), String> {
-		let json = Json::from_str(body).map_err(|e| format!("Invalid JSON returned: {:?}", e))?;
-		let obj = json.find_path(&["result", "ethusd"]).ok_or("USD price not found".to_owned())?;
-		let ethusd = match *obj {
-			Json::String(ref s) => FromStr::from_str(s).ok(),
-			_ => None,
-		}.ok_or("Unexpected price format.".to_owned())?;
-
-		(self.set_price)(PriceInfo {
-			ethusd: ethusd,
-		});
-		Ok(())
-	}
+pub struct Client {
+	api_endpoint: String,
+	fetch: FetchClient,
 }
 
-impl PriceInfo {
-	pub fn get<F: Fn(PriceInfo) + Sync + Send + 'static>(set_price: F) {
-		thread::spawn(move || {
-			let url = FromStr::from_str("http://api.etherscan.io/api?module=stats&action=ethprice")
-				.expect("string known to be a valid URL; qed");
-
-			if let Err(e) = Self::request(url, set_price) {
-				warn!("Failed to auto-update latest ETH price: {:?}", e);
-			}
-		});
+impl fmt::Debug for Client {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		write!(fmt, "Client {{ api_endpoint: {} }}", self.api_endpoint)
```

Comment:
maybe have the debug info indicate it's a price fetch client? We have another `Client` which is already very prolific throughout the codebase.